cmd/main.go

```diff
@@ -12,114 +13,125 @@ import (
+	coap-cli get channels/0bb5ba61-a66e-4972-bab6-26f19962678f/messages/subtopic --auth 1e1017e6-dee7-45b4-8a13-00e6afeb66eb -o `
 
-	code, err := parseCode(strings.ToUpper(os.Args[1]))
-	if err != nil {
-		log.Fatalf("Can't read request code: %s\n%s", err, helpCmd)
+	var putCmd = &cobra.Command{
+		Use:   "put <url>",
+		Short: "Perform a PUT request on a COAP resource",
+		Run: func(cmd *cobra.Command, args []string) {
+			if checkArgs(cmd, args) {
+				executeRequest(codes.PUT, args)
+			}
+			return
+		},
 	}
+	putCmd.Flags().StringVarP(&data, "data", "d", "", "Data")
 
-	if len(os.Args) < 3 {
-		log.Fatalf("CoAP URL must not be empty.\n%s", helpCmd)
+	var postCmd = &cobra.Command{
+		Use:   "post <url>",
+		Short: "Perform a POST request on a COAP resource",
+		Run: func(cmd *cobra.Command, args []string) {
+			if checkArgs(cmd, args) {
+				executeRequest(codes.POST, args)
+			}
+			return
+		},
 	}
-	path := os.Args[2]
-	if strings.HasPrefix(path, "-") {
-		log.Fatalf("Please enter a valid CoAP URL.\n%s", helpCmd)
+	postCmd.Flags().StringVarP(&data, "data", "d", "", "Data")
+	postCmd.Example = "coap-cli post channels/0bb5ba61-a66e-4972-bab6-26f19962678f/messages/subtopic --auth 1e1017e6-dee7-45b4-8a13-00e6afeb66eb -d \"hello world\""
+
+	var deleteCmd = &cobra.Command{
+		Use:   "delete <url>",
+		Short: "Perform a DELETE request on a COAP resource",
+		Run: func(cmd *cobra.Command, args []string) {
+			if checkArgs(cmd, args) {
+				executeRequest(codes.DELETE, args)
+			}
+			return
+		},
 	}
+	deleteCmd.Flags().StringVarP(&data, "data", "d", "", "Data")
+
+	rootCmd.AddCommand(getCmd, putCmd, postCmd, deleteCmd)
+	rootCmd.PersistentFlags().StringVarP(&host, "host", "H", "localhost", "Host")
+	rootCmd.PersistentFlags().StringVarP(&port, "port", "p", "5683", "Port")
+	rootCmd.PersistentFlags().StringVarP(&auth, "auth", "a", "", "Auth")
+	rootCmd.PersistentFlags().IntVarP(&contentFormat, "content-format", "c", 50, "Content format")
 
-	os.Args = os.Args[2:]
-	o := flag.Bool("o", false, "Observe")
-	h := flag.String("h", "localhost", "Host")
-	p := flag.String("p", "5683", "Port")
-	// Default type is JSON.
-	cf := flag.Int("cf", 50, "Content format")
-	d := flag.String("d", "", "Message data")
-	a := flag.String("auth", "", "Auth token")
-	flag.Parse()
-
-	client, err := coap.New(*h + ":" + *p)
+	if err := rootCmd.Execute(); err != nil {
+		log.Fatalf("Error executing command: %v", err)
+	}
+
+}
+
+func executeRequest(code codes.Code, args []string) {
+	client, err := coap.New(host + ":" + port)
 	if err != nil {
-		log.Fatal("Error creating client: ", err)
+		log.Fatalf("Error coap creating client: %v", err)
 	}
 	var opts coapmsg.Options
-	if a != nil {
-		opts = append(opts, coapmsg.Option{ID: coapmsg.URIQuery, Value: []byte(fmt.Sprintf("auth=%s", *a))})
+	if auth != "" {
+		opts = append(opts, coapmsg.Option{ID: coapmsg.URIQuery, Value: []byte("auth=" + auth)})
 	}
```

Comment:
Apart from `-a`,  It will be great if we have provision to pass option id and option value from CLI , like in `--options_id 33 --option_value some_value"`

Reply:
I like the idea, but it would be good to have some common options more user-friendly. While this is a generic client, we can have "sugars" for, say, observable and auth in the context of Magistrala. Otherwise, having `-O id,value` format is fine, as long as it's documented.